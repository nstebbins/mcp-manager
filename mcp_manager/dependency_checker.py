"""Module for checking if required dependencies are installed."""

import shutil
import subprocess
from typing import List, Tuple


def check_nodejs_npm() -> Tuple[bool, List[str]]:
    """
    Check if Node.js and npm are installed.

    Returns:
        Tuple of (all_installed: bool, missing_deps: List[str])
    """
    missing = []

    # Check node
    if not shutil.which("node"):
        missing.append("Node.js")

    # Check npm
    if not shutil.which("npm"):
        missing.append("npm")

    return len(missing) == 0, missing


def check_docker() -> Tuple[bool, List[str]]:
    """
    Check if Docker is installed and running.

    Returns:
        Tuple of (all_installed: bool, missing_deps: List[str])
    """
    missing = []

    # Check if docker is installed
    if not shutil.which("docker"):
        missing.append("Docker")
        return False, missing

    # Check if docker daemon is running
    try:
        subprocess.run(["docker", "info"], capture_output=True, check=True)
    except subprocess.CalledProcessError:
        missing.append("Docker daemon (not running)")

    return len(missing) == 0, missing


def check_dependencies(dependencies: List[str]) -> Tuple[bool, List[str]]:
    """
    Check if all required dependencies are installed.

    Args:
        dependencies: List of dependency names to check

    Returns:
        Tuple of (all_installed: bool, missing_deps: List[str])
    """
    missing = []

    for dep in dependencies:
        if dep in ["Node.js", "npm"]:
            installed, missing_deps = check_nodejs_npm()
            if not installed:
                missing.extend(missing_deps)
        elif dep == "Docker":
            installed, missing_deps = check_docker()
            if not installed:
                missing.extend(missing_deps)

    return len(missing) == 0, missing
