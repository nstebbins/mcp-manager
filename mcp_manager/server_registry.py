"""
Server registry containing information about available MCP servers.
Each server entry contains:
- description: A brief description of what the server does
- maintainer: Who maintains this server
- claude_config: The configuration needed for Claude settings file
- required_config: Any additional configuration needed
- dependencies: List of dependencies required
"""

import os
from typing import Any, Dict, List, Optional

MCP_SERVERS: Dict[str, Dict[str, Any]] = {
    "filesystem": {
        "description": "MCP server for filesystem operations",
        "maintainer": "Anthropic",
        "claude_config": {
            "command": "npx",
            "args": [
                "-y",
                "@modelcontextprotocol/server-filesystem",
                os.path.expanduser("~/Documents"),  # Default to user's Documents folder
            ],
        },
        "required_config": ["Allowed directory paths that the server can access"],
        "dependencies": ["Node.js", "npm"],
    }
}


def get_server_info(server_name: str) -> Optional[Dict[str, Any]]:
    """
    Get information about a specific server.

    Args:
        server_name: Name of the server to look up

    Returns:
        Server information dictionary if found, None otherwise
    """
    return MCP_SERVERS.get(server_name)


def search_servers(keyword: str) -> List[str]:
    """
    Search for servers matching the given keyword.

    Args:
        keyword: Search term to match against server names and descriptions

    Returns:
        List of matching server names
    """
    keyword = keyword.lower()
    matches = []

    for name, info in MCP_SERVERS.items():
        if keyword in name.lower() or keyword in info["description"].lower():
            matches.append(name)

    return matches


def get_claude_config(server_name: str) -> Optional[Dict[str, Any]]:
    """
    Get the Claude configuration for a specific server.

    Args:
        server_name: Name of the server to get config for

    Returns:
        Claude configuration dictionary if found, None otherwise
    """
    server_info = get_server_info(server_name)
    if not server_info:
        return None
    return server_info.get("claude_config")
