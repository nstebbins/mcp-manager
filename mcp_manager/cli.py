import json
import os
from enum import Enum
from pathlib import Path
from typing import Optional

import typer

from .server_registry import get_claude_config, get_server_info, search_servers

app = typer.Typer()


class ClientType(str, Enum):
    CURSOR = "cursor"
    CLAUDE = "claude"


class ScopeType(str, Enum):
    GLOBAL = "global"
    PROJECT = "project"


# Define options at module level
client_option = typer.Option(None, help="Client type (cursor or claude)")
scope_option = typer.Option(None, help="Installation scope (global or project)")


@app.command()
def search(keyword: str):
    """
    Search the online registry for servers matching the keyword.
    """
    matches = search_servers(keyword)
    if not matches:
        typer.echo(f"No servers found matching: {keyword}")
        return

    typer.echo(f"Found {len(matches)} matching servers:")
    for server_name in matches:
        info = get_server_info(server_name)
        typer.echo(f"\n{server_name}:")
        typer.echo(f"  Description: {info['description']}")
        typer.echo(f"  Maintainer: {info['maintainer']}")


@app.command()
def info(server_name: str):
    """
    Display detailed information about a specific server.
    """
    server_info = get_server_info(server_name)
    if not server_info:
        typer.echo(f"Server not found: {server_name}")
        return

    typer.echo(f"\nServer: {server_name}")
    typer.echo(f"Description: {server_info['description']}")
    typer.echo(f"Maintainer: {server_info['maintainer']}")

    if server_info["required_config"]:
        typer.echo("\nRequired configuration:")
        for config in server_info["required_config"]:
            typer.echo(f"- {config}")

    if server_info["dependencies"]:
        typer.echo("\nDependencies:")
        for dep in server_info["dependencies"]:
            typer.echo(f"- {dep}")


@app.command()
def install(
    server_name: str,
    client: Optional[ClientType] = client_option,
    scope: Optional[ScopeType] = scope_option,
):
    """
    Install a server with optional client type and scope specifications.
    """
    # Get server info and config
    server_info = get_server_info(server_name)
    if not server_info:
        typer.echo(f"Server not found: {server_name}")
        return

    # For now, we only support Claude installation
    if client and client != ClientType.CLAUDE:
        typer.echo("Currently only Claude installation is supported")
        return

    # Get Claude config
    claude_config = get_claude_config(server_name)
    if not claude_config:
        typer.echo(f"No Claude configuration available for server: {server_name}")
        return

    # Get Claude config file path
    config_dir = Path(os.path.expanduser("~/Library/Application Support/Claude"))
    config_file = config_dir / "claude_desktop_config.json"

    if not config_file.exists():
        typer.echo(f"Claude config file not found at: {config_file}")
        return

    try:
        # Read current config
        with open(config_file) as f:
            config = json.load(f)

        # Initialize mcpServers if it doesn't exist
        if "mcpServers" not in config:
            config["mcpServers"] = {}

        # Add or update server config
        config["mcpServers"][server_name] = claude_config

        # Write updated config
        with open(config_file, "w") as f:
            json.dump(config, f, indent=2)

        typer.echo(f"Successfully installed {server_name} for Claude")

    except Exception as e:
        typer.echo(f"Error updating Claude config: {str(e)}")
        return


def main():
    app()
