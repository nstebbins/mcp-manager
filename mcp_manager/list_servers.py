#!/usr/bin/env python3
"""
Command-line tool to list installed MCP servers.
"""

from server_registry import get_installed_servers


def main():
    installed_servers = get_installed_servers()

    if not installed_servers:
        print("No MCP servers found in Claude configuration.")
        return

    print("\nInstalled MCP Servers:")
    print("=" * 80)

    for server in installed_servers:
        print(f"\nServer: {server['name']}")
        print(f"Description: {server['description']}")
        print(f"Maintainer: {server['maintainer']}")
        print("-" * 40)


if __name__ == "__main__":
    main()
