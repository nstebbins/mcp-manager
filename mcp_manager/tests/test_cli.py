from unittest.mock import mock_open, patch

import pytest
from syrupy.assertion import SnapshotAssertion
from typer.testing import CliRunner

from mcp_manager.cli import app


@pytest.fixture
def runner() -> CliRunner:
    return CliRunner()


@pytest.fixture
def snapshot(snapshot: SnapshotAssertion) -> SnapshotAssertion:
    return snapshot


def test_search_command_with_match(runner: CliRunner, snapshot: SnapshotAssertion) -> None:
    """Test the search command with matching keyword"""
    result = runner.invoke(app, ["search", "file"])
    assert result.exit_code == 0
    assert result.output == snapshot


def test_search_command_no_match(runner: CliRunner, snapshot: SnapshotAssertion) -> None:
    """Test the search command with non-matching keyword"""
    result = runner.invoke(app, ["search", "nonexistent"])
    assert result.exit_code == 0
    assert result.output == snapshot


def test_info_command_existing(runner: CliRunner, snapshot: SnapshotAssertion) -> None:
    """Test the info command for existing server"""
    result = runner.invoke(app, ["info", "filesystem"])
    assert result.exit_code == 0
    assert result.output == snapshot


def test_info_command_nonexisting(runner: CliRunner, snapshot: SnapshotAssertion) -> None:
    """Test the info command for non-existing server"""
    result = runner.invoke(app, ["info", "nonexistent"])
    assert result.exit_code == 0
    assert result.output == snapshot


def test_install_command_nonexisting(runner: CliRunner, snapshot: SnapshotAssertion) -> None:
    """Test installing a non-existing server"""
    result = runner.invoke(app, ["install", "nonexistent"])
    assert result.exit_code == 0
    assert result.output == snapshot


def test_install_command_unsupported_client(runner: CliRunner, snapshot: SnapshotAssertion) -> None:
    """Test installing with unsupported client type"""
    result = runner.invoke(app, ["install", "filesystem", "--client", "cursor"])
    assert result.exit_code == 0
    assert result.output == snapshot


@patch("pathlib.Path.exists")
@patch("builtins.open", new_callable=mock_open, read_data='{"mcpServers": {}}')
def test_install_command_success(
    mock_file, mock_exists, runner: CliRunner, snapshot: SnapshotAssertion
) -> None:
    """Test successful server installation"""
    mock_exists.return_value = True

    result = runner.invoke(app, ["install", "filesystem", "--client", "claude"])
    assert result.exit_code == 0
    assert result.output == snapshot

    # Verify the file was written to
    mock_file.assert_called()
