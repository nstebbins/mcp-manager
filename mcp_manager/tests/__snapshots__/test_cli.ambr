# serializer version: 1
# name: test_info_command_existing
  '''

  Server: filesystem
  Description: MCP server for filesystem operations
  Maintainer: Anthropic

  Required configuration:
  - Allowed directory paths that the server can access

  Dependencies:
  - Node.js
  - npm

  '''
# ---
# name: test_info_command_nonexisting
  '''
  Server not found: nonexistent

  '''
# ---
# name: test_install_command_nonexisting
  '''
  Server not found: nonexistent

  '''
# ---
# name: test_install_command_success
  '''
  Successfully installed filesystem for Claude

  '''
# ---
# name: test_install_command_unsupported_client
  '''
  Currently only Claude installation is supported

  '''
# ---
# name: test_search_command_no_match
  '''
  No servers found matching: nonexistent

  '''
# ---
# name: test_search_command_with_match
  '''
  Found 1 matching servers:

  filesystem:
    Description: MCP server for filesystem operations
    Maintainer: Anthropic

  '''
# ---
